kratos/tensor_adaptors/tensor_adaptor.h

```diff
@@ -0,0 +1,421 @@
+class KRATOS_API(KRATOS_CORE) TensorAdaptor {
+protected:
+    ///@name Class definitions
+    ///@{
+
+    /**
+     * @class Storage
+     * @brief Manages the storage and lifetime of tensor adaptor data associated with various Kratos ModelPart containers.
+     *
+     * The Storage class encapsulates a pointer to a container (such as nodes, elements, conditions, etc.) and manages
+     * the associated tensor data. It provides mechanisms for copying, moving, and viewing the internal data, as well as
+     * querying the shape and size of the tensor. Reference counting is implemented for safe memory management using
+     * intrusive pointers.
+     *
+     * @tparam TDataType The type of the data stored in the tensor adaptor.
+     */
+    class Storage
+    {
+    public:
+        ///@name Type definitions
+        ///@{
+
+        KRATOS_CLASS_INTRUSIVE_POINTER_DEFINITION(Storage);
+
+        using ContainerPointerType = std::variant<
+                                            ModelPart::DofsArrayType::Pointer,
+                                            ModelPart::NodesContainerType::Pointer,
+                                            ModelPart::ConditionsContainerType::Pointer,
+                                            ModelPart::ElementsContainerType::Pointer,
+                                            ModelPart::PropertiesContainerType::Pointer,
+                                            ModelPart::MasterSlaveConstraintContainerType::Pointer,
+                                            ModelPart::GeometryContainerType::Pointer
+                                        >;
+
+        ///@}
+        ///@name Life cycle
+        ///@{
+
+        Storage(
+            ContainerPointerType pContainer,
+            const DenseVector<unsigned int>& rShape);
+
+        /**
+         * @brief Destroy the Tensor Adaptor storage
+         * @details This method destroys the TensorAdaptorStorage and
+         *          the internal data as well if it is still owned by the TensorAdaptorStorage.
+         *
+         */
+        ~Storage();
+
+        ///@}
+        ///@name Public operations
+        ///@{
+
+        /**
+         * @brief Creates a copy of the current storage object.
+         * @details This method returns a pointer to a new Storage object that is having copied data and copied
+         *          pointer to the container.
+         * @return Storage::Pointer A pointer to the copied Storage object.
+         */
+        Storage::Pointer Copy() const;
+
+        /**
+         * @brief Moves the internal data.
+         * @warning The TensorAdaptor should not be used after the move is called.
+         *          The management of the data should be carried out by the owner of the
+         *          returned Kratos::span. Otherwise, there will be memory leaks.
+         * @throws If the internal data is already moved.
+         * @return Kratos::span<TDataType>  Returns a span containing the internal data.
```

Comment:
Doxygen is type-aware.

```suggestion
         * @return Returns a span containing the internal data.
```